lighthouse-core/lib/traces/tracing-processor.js

```diff
@@ -241,23 +188,45 @@ class TraceProcessor {
   }
 
   /**
-   * Provides the top level events on the main thread.
-   * @param {!traceviewer.Model} model
-   * @param {{traceEvents: !Array<!Object>}} trace
+   * Provides the top level events on the main thread with timestamps in ms relative to navigation
+   * start.
+   * @param {!TraceOfTabArtifact} tabTrace
    * @param {number=} startTime Optional start time (in ms) of range of interest. Defaults to trace start.
    * @param {number=} endTime Optional end time (in ms) of range of interest. Defaults to trace end.
    * @return {!Array<{start: number, end: number, duration: number}>}
    */
-  static getMainThreadTopLevelEvents(model, trace, startTime = -Infinity, endTime = Infinity) {
-    // Find the main thread via the first TracingStartedInPage event in the trace
-    const startEvent = trace.traceEvents.find(event => event.name === 'TracingStartedInPage');
-    const mainThread = TraceProcessor._findMainThreadFromIds(model, startEvent.pid, startEvent.tid);
-
-    return mainThread.sliceGroup.slices.filter(slice => {
-      return slice.title === SCHEDULABLE_TASK_TITLE &&
-          slice.end > startTime &&
-          slice.start < endTime;
-    });
+  static getMainThreadTopLevelEvents(tabTrace, startTime = -Infinity, endTime = Infinity) {
+    let lastEvt = {start: -Infinity, end: -Infinity};
+    const topLevelEvents = [];
+    // note: processEvents is already sorted by event start
+    for (const event of tabTrace.processEvents) {
+      if (event.name !== SCHEDULABLE_TASK_TITLE || !event.dur) continue;
+
+      const start = (event.ts - tabTrace.navigationStartEvt.ts) / 1000;
+      const end = (event.ts + event.dur - tabTrace.navigationStartEvt.ts) / 1000;
+      if (start > endTime || end < startTime) continue;
+
+      const currentEvt = {
+        start,
+        end,
+        duration: event.dur / 1000,
+        events: [event],
+      };
+
+      if (currentEvt.end < lastEvt.end) {
+        // The current event is completely inside the previous event, add to previous
+        lastEvt.events.push(event);
+      } else if (currentEvt.start < lastEvt.end) {
+        // The current event overlaps with the previous event, extend the duration and add to previous
+        lastEvt.events.push(event);
+        lastEvt.end = currentEvt.end;
+        lastEvt.duration = currentEvt.end - currentEvt.start;
```

Comment:
should this be `lastEvt.duration = currentEvt.end - lastEvt.start;` (or `currentEvt.end - currentEvt.start`) if extending duration by the end of `currentEvt` (but keeping the same `start`)?